### ISSUE TITLE
ec2_elb not able to register previously unregistered instance

### ISSUE THREAD

**Issue by conradlee** [ansible/ansible#4902]
Summary: currently `ec2_elb` can only register an instance on an elastic load balancer if that instance has a status on that load balancer.  The module should be able to register instances with a load balancer that has no status information on that instance.

Here's why it currently can't do so.  Here's the current `register` method in `library/cloud/ec2_elb`

``` python
    def register(self, wait, enable_availability_zone):
        """Register the instance for all ELBs and wait for the ELB
        to report the instance in-service"""
        for lb in self.lbs:
            initial_state = lb.get_instance_health([self.instance_id])[0]
            if enable_availability_zone:
                self._enable_availailability_zone(lb)
            lb.register_instances([self.instance_id])
            if wait:
                self._await_elb_instance_state(lb, 'InService', initial_state)
            else:
                # We cannot assume no change was made if we don't wait
                # to find out
                self.changed = True
```

Check out the line that get's the initial state: `lb.get_instance_health([self.instance_id])[0]`.  This asks the load balancer to get the instance's current health.  If the instance was never previously registered on the load balancer, then this will cause boto to raise a `boto.exception.BotoServerError`.

The conceptual problem is that the `register` method needs to have an initial state because later on it calls the `_await_instance_state` method, which requires an initial state.  But the instance has no initial state because the load balancer doesn't know anything about it.

**Comment by conradlee** [comment]
My proposed solution is to allow the method `_await_elb_instance_state` to take a null initial state by passing `None` as the value for the initial state.

This involves changing one line of code in the `_await_elb_instance_state` method:

``` python
if instance_state.state != initial_state.state:
```

becomes

``` python
if initial_state is None or instance_state.state != initial_state.state:
```

Thus, the modified method looks like:

``` python
    def _await_elb_instance_state(self, lb, awaited_state, initial_state):
        """Wait for an ELB to change state
        lb: load balancer
        awaited_state : state to poll for (string)"""
        while True:
            instance_state = lb.get_instance_health([self.instance_id])[0]
            if instance_state.state == awaited_state:
                # Check the current state agains the initial state, and only set
                # changed if they are different.
                if initial_state is None or instance_state.state != initial_state.state:
                    self.changed = True
                break
            elif (awaited_state == 'InService'
                  and instance_state.reason_code == "Instance"):
                # If the reason_code for the instance being out of service is
                # "Instance" this indicates a failure state, e.g. the instance
                # has failed a health check or the ELB does not have the
                # instance's availabilty zone enabled. The exact reason why is
                # described in InstantState.description.
                msg = ("The instance %s could not be put in service on %s."
                       " Reason: %s")
                self.module.fail_json(msg=msg % (self.instance_id,
                                                 lb,
                                                 instance_state.description))
            else:
                time.sleep(1)
```

Then, if we find that the elastic load balancer has no state information on the instance, we can just use `None` as the initial state.  This involves changing a couple of lines of codes in the `register` method.

``` python
initial_state = lb.get_instance_health([self.instance_id])[0]
```

becomes

``` python
try:
    initial_state = lb.get_instance_health([self.instance_id])[0]
except boto.exception.BotoServerError as e:
    initial_state = None
```

Thus, The new `register` method would thus look like:

``` python
    def register(self, wait, enable_availability_zone):
        """Register the instance for all ELBs and wait for the ELB
        to report the instance in-service"""
        for lb in self.lbs:
            try:
                initial_state = lb.get_instance_health([self.instance_id])[0]
            except boto.exception.BotoServerError as e:
                initial_state = None
            if enable_availability_zone:
                self._enable_availailability_zone(lb)
            lb.register_instances([self.instance_id])
            if wait:
                self._await_elb_instance_state(lb, 'InService', initial_state)
            else:
                # We cannot assume no change was made if we don't wait
                # to find out
                self.changed = True
```

**Comment by jarv** [comment]
This looks like it changed with the 1.4 refactor in https://github.com/ansible/ansible/pull/4112 , the proposal seems fine with me, @jsdalton can you weigh in?

**Comment by jsdalton** [comment]
@conradlee Cool, good catch. I concur with your assessment and your proposal 100%. I'd suggest maybe throwing your code into a pull request and submitting it -- that might end up being a quicker pathway to getting it fixed. But if a committer wants to just fix the issue that'd be great too.

My only comment was that we ask for the initial state during the `deregister` process as well...and I wonder if it makes sense to wrap the call there in a try / except block as well? I'm thinking no, because I can't think of a reason why it would fail in that case (and if it does fail it's probably for a different reason that should not be caught).

**Comment by conradlee** [comment]
@jsdalton In the case of `deregister` the error that comes up is appropriate.  My own fork of ansible includes other changes, so I'm  reluctant to do a pull request.  [Here's a link](https://github.com/conradlee/ansible/commit/b39265b1445cc4a0ec4abfe301a65b9505504858#diff-65143e5debe6a07b9e81d5d14fc8abbb) to a commit that fixes the bug in the way I have proposed.

I've also found a bit more undesirable behavior in `ec2_elb` that also has a simple fix.  I'm not sure whether to mention it here or in a new github issue, but as I've got the relevant devs' attention here already I'll just do it here.  Basically, upon registering a new instance, `ec2_elb` raises an error a when the instance passes into the "pending" state.  This should not happen, because we should expect the instance to spend a little time in the pending state before entering the 'InService' state.  [Here's a link](https://github.com/conradlee/ansible/commit/168c195053622c6218d9edfd04d7c3c9e0f7606e#diff-65143e5debe6a07b9e81d5d14fc8abbb) to a commit that fixes this bug.

**Comment by mpdehaan** [comment]
Thanks for looking into this folks -- Please submit a pull request.

**Comment by jsdalton** [comment]
Issue #5076 is actually a dupe of this one. I actually forgot about this issue and submitted pull request on that issue which addresses it. The pull request also addresses a few other minor issues that were similar to this one (e.g. it wasn't handling the pending state on initial registration properly).

@conradlee - Feel free to take a look at that other issue / pull request, and you might want to close this one if you think the pull request suffices. It's a pretty similar approach to the one you suggested in the comments here.

**Comment by jctanner** [comment]
@conradlee https://github.com/ansible/ansible/pull/5232 was merged in. Please let us know if it does not resolve your issue.

**Comment by hkariti** [comment]
Seems like this is happening again. The following was added to register():

```
                initial_state = self._get_instance_health(lb)
                tries = 1
                while True:
                    if initial_state:
                        break
                    time.sleep(1)
                    tries += 1
                    # FIXME: this should be configurable, but since it didn't
                    #        wait at all before this is at least better
                    if tries > 10:
                        self.module.fail_json(msg='failed to find the initial state of the load balancer')
```

Which fails, because _get_instance_health returns None when the instance hasn't been registered before, which prevents the loop from breaking.

**Comment by esigler** [comment]
I've also run into this issue, as of Ansible 1.5.5.

**Comment by jimi-c** [comment]
I've added a new test to the ec2_elb integration test for this, and it has passed, so it appears that the issue is resolved in the development branch and will be included in 1.6.0 when it is released in the near future. Thanks!
